nexus/db-queries/src/db/datastore/instance.rs

```diff
@@ -818,35 +822,57 @@ impl DataStore {
+            .inner_join(vmm_dsl::vmm.on(vmm_dsl::sled_id.eq(sled_dsl::id)));
+        // Next, paginate the results, ordering by the sled ID first, so that we
+        // list all VMMs on a sled before moving on to the next one, and then by
+        // the VMM ID.
+        //
+        // Note that we must add the `paginated_multicolumn` wrapper
+        // at this point in query construction, because here, the selection
+        // contains both `sled_dsl::id` and `vmm_dsl::id` columns, but it does
+        // *not* have anything that makes it no longer implement
+        // `diesel::QuerySource`, which the `paginated_multicolumn` function
+        // requires. This ordering doesn't actually matter when it comes to the
+        // generated SQL, which should be equivalent no matter how we construct
+        // the query, but it *does* matter for satisfying Diesel's trait
+        // constraints.
+        let query = paginated_multicolumn(
+            query,
+            (sled_dsl::id, vmm_dsl::id),
+            pagparams,
+        );
+        // Filter out sleds that aren't in service, and VMMs that aren't
+        // incarnated on a sled.
+        let query = query
             .filter(sled_dsl::time_deleted.is_null())
             .sled_filter(SledFilter::InService)
-            .inner_join(
-                vmm_dsl::vmm
-                    .on(vmm_dsl::sled_id
-                        .eq(sled_dsl::id)
-                        .and(vmm_dsl::time_deleted.is_null())
-                        // Ignore instances which are in states that are not
-                        // known to exist on a sled. Since this query drives
-                        // instance-watcher health checking, it is not necessary
-                        // to perform health checks for VMMs that don't actually
-                        // exist in real life.
-                        .and(
-                            vmm_dsl::state.ne_all(VmmState::NONEXISTENT_STATES),
-                        ))
-                    .inner_join(
-                        instance_dsl::instance
-                            .on(instance_dsl::id
-                                .eq(vmm_dsl::instance_id)
-                                .and(instance_dsl::time_deleted.is_null()))
-                            .inner_join(
-                                project_dsl::project.on(project_dsl::id
-                                    .eq(instance_dsl::project_id)
-                                    .and(project_dsl::time_deleted.is_null())),
-                            ),
-                    ),
-            )
-            .sled_filter(SledFilter::InService)
+            .filter(vmm_dsl::time_deleted.is_null())
+            // Ignore VMMs which are in states that are not known to exist on a
+            // sled. Since this query drives instance-watcher health checking,
+            // it is not necessary to perform health checks for VMMs that don't
+            // actually exist in real life.
```

Comment:
This is probably tangential to all of this work, but do we have anything that notices a case where a VM is unexpectedly running on a sled?

Reply:
Hmm, the `instance-watcher` task will not do that, since it's asking if the sled-agents actually have the instances they're _believed_ to have, rather than asking them "hey, what instances do you have?". 

I think there may be a code path in `cpapi_instances_put` that handles the case where a sled-agent has sent nexus an update regarding a VMM that Nexus doesn't believe exists, but off the top of my head, I don't think that code path does anything besides returning an error to the sled-agent...